src/app-layer-htp-file.c

```diff
@@ -225,34 +261,122 @@ int HTPParseContentRange(bstr * rawvalue, HtpContentRange *range)
-    if (retval == -1) {
-        SCLogDebug("set range failed");
+    uint8_t *keyurl;
+    size_t keylen;
+    if (tx->request_hostname != NULL) {
+        keylen = bstr_len(tx->request_hostname) + filename_len + 1;
+        keyurl = SCMalloc(keylen);
+        if (keyurl == NULL) {
+            SCReturnInt(-1);
+        }
+        memcpy(keyurl, bstr_ptr(tx->request_hostname), bstr_len(tx->request_hostname));
+        memcpy(keyurl + bstr_len(tx->request_hostname), filename, filename_len);
+        keyurl[keylen - 1] = 0;
+    } else {
+        // do not reassemble file without host info
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
+    }
+    HttpRangeContainerFile *file_range_container =
+            HttpRangeContainerUrlGet(keyurl, keylen, &s->f->lastts);
+    SCFree(keyurl);
+    if (file_range_container == NULL) {
+        // probably reached memcap
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
+    }
+    DEBUG_VALIDATE_BUG_ON(s->file_range);
+    s->file_range = HttpRangeOpenFile(file_range_container, crparsed.start, crparsed.end,
+            crparsed.size, &s->cfg->response.sbcfg, filename, filename_len, flags, data, data_len);
+    SCLogDebug("s->file_range == %p", s->file_range);
+    if (s->file_range == NULL) {
+        SCLogDebug("s->file_range == NULL");
+        THashDecrUsecnt(file_range_container->hdata);
+        DEBUG_VALIDATE_BUG_ON(
+                SC_ATOMIC_GET(file_range_container->hdata->use_cnt) > (uint32_t)INT_MAX);
+        THashDataUnlock(file_range_container->hdata);
+
+        // probably reached memcap
+        return HTPFileOpen(
+                s, filename, (uint32_t)filename_len, data, data_len, txid, STREAM_TOCLIENT);
+        /* in some cases we don't take a reference, so decr use cnt */
+    } else if (s->file_range->container == NULL) {
+        THashDecrUsecnt(file_range_container->hdata);
+    } else {
+        SCLogDebug("container %p use_cnt %u", s->file_range,
+                SC_ATOMIC_GET(s->file_range->container->hdata->use_cnt));
+        DEBUG_VALIDATE_BUG_ON(
+                SC_ATOMIC_GET(s->file_range->container->hdata->use_cnt) > (uint32_t)INT_MAX);
+    }
+
+    /* we're done, so unlock. But since we have a reference in s->file_range keep use_cnt. */
+    THashDataUnlock(file_range_container->hdata);
+    SCReturnInt(0);
+}
+
+static void HTPFileStoreChunkHandleRange(
+        HttpRangeContainerBlock *c, const uint8_t *data, uint32_t data_len)
+{
+    if (c->container) {
+        THashDataLock(c->container->hdata);
+        DEBUG_VALIDATE_BUG_ON(SC_ATOMIC_GET(c->container->hdata->use_cnt) == 0);
+        if (HttpRangeAppendData(c, data, data_len) < 0) {
+            SCLogDebug("Failed to append data");
+        }
+        DEBUG_VALIDATE_BUG_ON(SC_ATOMIC_GET(c->container->hdata->use_cnt) > (uint32_t)INT_MAX);
```

Comment:
This `DEBUG_VALIDATE_BUG_ON` still feels redundant as https://github.com/OISF/suricata/pull/6298#discussion_r694120958